pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,473 @@
+		cfg, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return newPCCloud(cfg)
+	})
+}
+
+func getVMIDbyNodename(project string, nodeName string) (string, error) {
+	vmList, err := Esxclient.Projects.GetVMs(project, nil)
+	if err != nil {
+		glog.V(2).Info("Photon Cloud Provider: Failed to GetVMs from project %s with nodeName %s", project, nodeName)
+		return "", err
+	}
+
+	for _, vm := range vmList.Items {
+		if vm.Name == nodeName {
+			// TODO: check vm state
+			return vm.ID, nil
+		}
+	}
+
+	return "", fmt.Errorf("Photon Cloud Provider: no matching VM is found with name %s", nodeName)
+}
+
+func getVMIDbyIP(project string, IPAddress string) (string, error) {
+	vmList, err := Esxclient.Projects.GetVMs(project, nil)
+	if err != nil {
+		glog.V(2).Info("Photon Cloud Provider: Failed to GetVMs from project %s, IP %s", project, IPAddress)
+		return "", err
+	}
+
+	for _, vm := range vmList.Items {
+		task, err := Esxclient.VMs.GetNetworks(vm.ID)
+		if err != nil {
+			glog.V(2).Info("Photon Cloud Provider: GetNetworks failed for vm.ID %s")
+		} else {
+			task, err = Esxclient.Tasks.Wait(task.ID)
+			if err != nil {
+				glog.V(2).Info("Photon Cloud Provider: , Wait task for GetNetworks failed for vm.ID %s")
+			} else {
+				networkConnections := task.ResourceProperties.(map[string]interface{})
+				networks := networkConnections["networkConnections"].([]interface{})
+				for _, nt := range networks {
+					network := nt.(map[string]interface{})
+					if val, ok := network["ipAddress"]; ok && val != nil {
+						ipAddr := val.(string)
+						if ipAddr == IPAddress {
+							return vm.ID, nil
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("Photon Cloud Provider: no matching VM is found with IP %s", IPAddress)
+}
+
+func getProjIDbyName(tenantName, projName string) (string, error) {
+	tenants, err := Esxclient.Tenants.GetAll()
+	if err != nil {
+		return "", err
+	}
+
+	for _, tenant := range tenants.Items {
+		if tenant.Name == tenantName {
+			projects, err := Esxclient.Tenants.GetProjects(tenant.ID, nil)
+			if err != nil {
+				return "", err
+			}
+
+			for _, project := range projects.Items {
+				if project.Name == projName {
+					return project.ID, nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("Photon Cloud Provider: no matching tenant/project name is found with %s/%s", tenantName, projName)
+}
+
+func newPCCloud(cfg PCConfig) (*PCCloud, error) {
+	if len(cfg.Global.CloudTarget) == 0 {
+		return nil, errors.New("Invalid Photon Controller endpoint")
+	}
+
+	//TODO: add handling of certification enabled situation
+	options := &photon.ClientOptions{
+		IgnoreCertificate: cfg.Global.IgnoreCertificate,
+	}
+
+	Esxclient = photon.NewClient(cfg.Global.CloudTarget, options, logger)
+	status, err := Esxclient.Status.Get()
+	if err != nil {
+		return nil, errors.New("Photon Controller new client creation failed")
```

Comment:
Assuming Esxclient returns some meanigful status and error. It would be helpful to include status and/or error in error message so that user can resolve the issue.